space_packet_parser/xtce/definitions.py

```diff
@@ -1,47 +1,162 @@
-        '{{{xtce}}}RelativeTimeParameterType': parameters.RelativeTimeParameterType,
-    }
+class XtcePacketDefinition(common.AttrComparable):
+    """Object representation of the XTCE definition of a CCSDS packet object"""
 
+    # TODO: Allow user to specify the XML schema instance for the XTCE XSD as well
+    #  e.g.
+    #                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    #                   xsi:schemaLocation="http://www.omg.org/spec/XTCE/20180204/SpaceSystem.xsd"
+    #  This will require an additional namespace dict entry for xsi (in this example)
     def __init__(
             self,
+            sequence_container_list: Optional[list[containers.SequenceContainer]] = None,
+            *,
+            ns: dict = XTCE_NSMAP,
+            xtce_schema_uri: str = XTCE_URI,
+            root_container_name: Optional[str] = DEFAULT_ROOT_CONTAINER,
+            space_system_name: Optional[str] = None,
+            xtce_version: str = "1.0",
+            date: str = None,
+            author: Optional[str] = None
+    ):
+        f"""
+
+        Parameters
+        ----------
+        sequence_container_list : Optional[list[space_packet_parser.containers.SequenceContainer]]
+            List of SequenceContainer objects, containing entry lists of Parameter objects, which contain their
+            ParameterTypes. This is effectively the entire XTCE document in one list of objects.
+        ns : dict
+            XML namespace, expected as a single entry dictionary with the key being the namespace label and the
+            value being the namespace URI. Default {XTCE_NSMAP}.
+        xtce_schema_uri : str
+            XTCE schema URI. This is required in addition to the ns in order to know _which_ ns key corresponds to XTCE.
+            Default {XTCE_URI}. When creating XTCE elements from objects, this is the namespace that is used.
+        root_container_name : Optional[str]
+            Name of root sequence container (where to start parsing)
+        space_system_name : Optional[str]
+            Name of space system to encode in XML when serializing.
+        xtce_version : str
+            Default "1.0"
+        date: Optional[str]
+            Optional header date string.
+        author : Optional[str]
+            Optional author name to include in XML when serializing.
+        """
+        self._parameter_type_cache = {}
+        self._parameter_cache = {}
+        self._sequence_container_cache = {}
+
+        # Populate the three caches for easy lookup later.
+        if sequence_container_list:
+            for sc in sequence_container_list:
+                self._parameter_type_cache.update({p.parameter_type.name: p.parameter_type for p in sc.entry_list})
+                self._parameter_cache.update({p.name: p for p in sc.entry_list})
+            self._sequence_container_cache.update({sc.name: sc for sc in sequence_container_list})
```

Comment:
This is where I'm still not fully following everything.

1. Rather than calling it a private cache and having the property indirection, why can't we just say `self.parameter_types: dict[str, ParameterType]  = {}` and remove the properties.
2. Why do we update/overwrite entries in these dictionaries? This seems like it could lead to potential issues down the line if someone were to pass in "FILL" as a uint4 in one container, but a "FILL" as uint6 in another. I'm not sure I have a great appreciation for when someone will create the definition this way though which I think might be where my confusion is coming from. I'm used to thinking of just coming straight from the XML which is now a separate method, so there are kind of two different init's in a sense.

Reply:
1. Ah! I understand now. I don't mind making them public. I think initially I was planning to keep them private, but then I went and added property accessors when I should have just made them public. I'll do this in a separate PR though.
2. You're right that we need to re-evaluate how these dicts are filled and updated (both when parsing a document and initializing directly with python objects). We are creating multiple objects for the same Parameter and then only storing the last one. I don't want to tackle this now but I do want to tackle it. It's working, it's just not clean.